libs/dmlf/examples/word2vec_multiprocess_distributed_learning/main.cpp

```diff
@@ -0,0 +1,143 @@
+#include <string>
+#include <thread>
+#include <vector>
+
+using namespace fetch::ml::ops;
+using namespace fetch::ml::layers;
+using namespace fetch::ml;
+using namespace fetch::ml::dataloaders;
+using namespace fetch::ml::distributed_learning;
+
+using DataType         = fetch::fixed_point::FixedPoint<32, 32>;
+using TensorType       = fetch::math::Tensor<DataType>;
+using VectorTensorType = std::vector<TensorType>;
+using SizeType         = typename TensorType::SizeType;
+
+int main(int argc, char **argv)
+{
+  if (argc != 6)
+  {
+    std::cout
+        << "Usage : " << argv[0]
+        << " PATH/TO/text8 analogies_test_file output_csv_file networker_config instance_number"
+        << std::endl;
+    return 1;
+  }
+
+  W2VTrainingParams<DataType> client_params;
+
+  // Command line parameters
+  std::string train_file            = argv[1];
+  client_params.analogies_test_file = argv[2];
+  std::string output_csv_file       = argv[3];
+  std::string config                = std::string(argv[4]);
+  int         instance_number       = std::atoi(argv[5]);
+
+  // Distributed learning parameters:
+  SizeType number_of_rounds = 1000;
+  SizeType number_of_peers  = 3;
+  // have been processed in total by the clients
+
+  client_params.batch_size    = 10000;
+  client_params.learning_rate = static_cast<DataType>(.001f);
+  client_params.max_updates   = 100;  // Round ends after this number of batches
+
+  // Word2Vec parameters:
+  client_params.vocab_file           = "/tmp/vocab.txt";
+  client_params.negative_sample_size = 5;  // number of negative sample per word-context pair
+  client_params.window_size          = 5;  // window size for context sampling
+  client_params.freq_thresh          = DataType{0.001f};  // frequency threshold for subsampling
+  client_params.min_count            = 5;                 // infrequent word removal threshold
+  client_params.embedding_size       = 100;               // dimension of embedding vec
+  client_params.starting_learning_rate_per_sample =
+      DataType{0.0025f};  // these are the learning rates we have for each sample
+  client_params.test_frequency = 1000;
+
+  client_params.k     = 20;       // how many nearest neighbours to compare against
+  client_params.word0 = "three";  // test word to consider
+  client_params.word1 = "king";
+  client_params.word2 = "queen";
+  client_params.word3 = "father";
+
+  // calc the true starting learning rate
+  client_params.starting_learning_rate = static_cast<DataType>(client_params.batch_size) *
+                                         client_params.starting_learning_rate_per_sample;
+  client_params.ending_learning_rate = static_cast<DataType>(client_params.batch_size) *
+                                       client_params.ending_learning_rate_per_sample;
+  client_params.learning_rate_param.starting_learning_rate = client_params.starting_learning_rate;
+  client_params.learning_rate_param.ending_learning_rate   = client_params.ending_learning_rate;
```

Comment:
presumably some of this is copy paste from the other word2vec example? We should try to avoid repeated code. Can you unify these together into some kind of utility/config function call somewhere?